R/call_utils.R

```diff
@@ -49,108 +21,96 @@ call_check_parse_varname <- function(varname) {
-#' teal.slice:::call_condition_choice("SEX", choices = factor(c("F", "M")))
-#' teal.slice:::call_condition_choice("x$SEX", choices = Sys.Date())
-#' teal.slice:::call_condition_choice("trunc(x$SEX)", choices = Sys.time())
+#' \dontrun{
+#' call_condition_choice("SEX", choices = c(1, 2))
+#' call_condition_choice("SEX", choices = "F")
+#' call_condition_choice("SEX", choices = c("F", "M"))
+#' call_condition_choice("x$SEX", choices = Sys.Date())
+#' call_condition_choice("x$SEX", choices = Sys.time())
+#' }
 #' @return a `call`
 #' @keywords internal
 call_condition_choice <- function(varname, choices) {
-  varname <- call_check_parse_varname(varname)
-
-  if (is.factor(choices)) {
-    choices <- as.character(choices)
-  } else if (inherits(choices, "Date")) {
-    choices <- format(choices)
-  } else if (inherits(choices, c("POSIXct", "POSIXlt"))) {
-    choices <- format(choices)
-  }
-
+  checkmate::assert_string(varname)
+  checkmate::assert_multi_class(
+    choices,
+    c("integer", "numeric", "character", "factor", "Date", "POSIXct", "POSIXlt")
+  )
 
-  if (length(choices) == 1) {
-    call("==", varname, choices)
+  if (is.numeric(choices)) {
+    if (length(choices) == 1) {
+      sprintf("%s == %s", varname, choices)
+    } else {
+      sprintf("%s %%in%% c(%s)", varname, toString(choices))
+    }
   } else {
-    c_call <- do.call(
-      "call",
-      append(list("c"), choices)
-    )
-    # c_call needed because it needs to be vector call
-    # instead of vector. SummarizedExperiment.subset
-    # handles only vector calls
-    call("%in%", varname, c_call)
+    if (length(choices) == 1) {
+      sprintf("%s == \"%s\"", varname, choices)
+    } else {
+      sprintf("%s %%in%% c(%s)", varname, toString(sprintf("\"%s\"", choices)))
+    }
   }
 }
 
 #' `numeric` range condition call
 #'
 #' Compose `numeric` range condition call from inputs
 #'
-#' @param varname (`name` or `character(1)`)\cr
+#' @param varname (`character(1)`)\cr
 #'   name of the variable
 #'
 #' @param range (`numeric(2)`)\cr
 #'   range of the variable
 #'
 #' @return call
 #' @examples
-#' teal.slice:::call_condition_range("AGE", range = c(1, 2))
-#' teal.slice:::call_condition_range(as.name("AGE"), range = c(-1.2, 2.1))
-#' teal.slice:::call_condition_range(
-#'   teal.slice:::call_extract_list("ADSL", "AGE"),
+#' \dontrun{
+#' call_condition_range("AGE", range = c(1, 2))
+#' call_condition_range(
+#'   call_extract_list("ADSL", "AGE"),
 #'   range = c(-1.2, 2.1)
 #' )
-#' @return a `call`
+#' }
+#' @return a character string
 #' @keywords internal
 call_condition_range <- function(varname, range) {
+  checkmate::assert_string(varname)
   checkmate::assert_numeric(range, len = 2, sorted = TRUE)
-  varname <- call_check_parse_varname(varname)
-  call(
-    "&",
-    call(">=", varname, range[1]),
-    call("<=", varname, range[2])
-  )
+
+  sprintf("%1$s >= %2$.4f & %1$s <= %3$.4f", varname, range[1], range[2])
```

Comment:
I think we should change this precision to 10, to follow what we have in the slider. It is still possible that slider initializes with double having 10 digits precision
```suggestion
  sprintf("%1$s >= %2$.10f & %1$s <= %3$.10f", varname, range[1], range[2])
```

```r
devtools::load_all("teal.slice")
library(scda)

ADSL <- synthetic_cdisc_data("latest")$adsl
ADSL$logical <- sample(c(TRUE, FALSE), size = nrow(ADSL), replace = TRUE)
ADSL$numeric <- seq(0, by = 1e-8, length.out = nrow(ADSL))

ADTTE <- synthetic_cdisc_data("latest")$adtte
ADRS <- synthetic_cdisc_data("latest")$adrs


funny_module <- function (label = "Filter states", datanames = "all") {
  checkmate::assert_string(label)
  module(
    label = label,
    filters = datanames,
    ui = function(id, ...) {
      ns <- NS(id)
      div(
        h2("The following filter calls are generated:"),
        verbatimTextOutput(ns("filter_states")),
        verbatimTextOutput(ns("filter_calls")),
        actionButton(ns("reset"), "reset_to_default")
      )
    },
    server = function(input, output, session, data, filter_panel_api) {
      checkmate::assert_class(data, "tdata")
      observeEvent(input$reset, set_filter_state(filter_panel_api, default_filters))
      output$filter_states <-  renderText({
        logger::log_trace("rendering text1")
        filter_panel_api |> get_filter_state() |> yaml::as.yaml()
      })
      output$filter_calls <- renderText({
        logger::log_trace("rendering text2")
        attr(data, "code")()
      })
    }
  )
}

data <- cdisc_data(
  cdisc_dataset("ADSL", ADSL),
  cdisc_dataset("ADTTE", ADTTE),
  cdisc_dataset("ADRS", ADRS)
)

app <- init(
  data = data,
  modules = list(funny_module(), funny_module("module2")),
  filter = list(ADSL = list(numeric = list()))
)


runApp(app)

```

![image](https://user-images.githubusercontent.com/6959016/219629214-7a97cd81-95b0-4aa6-87fe-8cea5a62b3c5.png)

Reply:
> EDIT: shall I increase precision? Or perhaps we want to define this differently, say %.ng?

I'm indifferent about `%g` and `%f`, I think `%f` is better than %g to present in the code for a lower precision numbers (most of the cases)

Reply:
We also use `toString` for numbers, which seems to round to the 14th decimal. Shall I round numerics to the 10th before passing to `toString` as well?

Reply:
I set all number-to-string convertions to round the 10th decimal. Seems a little better now:
![image](https://user-images.githubusercontent.com/114988527/220045401-05a7494a-e5f3-451a-ad05-17fe17601ead.png)